### PR TITLE
SE - Nullable: Infer constraints for `HasValue` property

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Part of #6812

Set `BoolConstraint` on `.HasValue` from known `ObjectConstraint`, if present.
Do not set `NotNull` for the symbol itself, as `something.HasValue` doesn't mean it cannot be null in this case.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs [hunk 1]
```diff
@@ -59,4 +59,30 @@ public void Nullable_Value_ReadsConstraintsFromInstance()
...
+            arg = true;
+            hasValue = arg.HasValue;
+            Tag("HasValueAfterTrue", hasValue);
+            Tag("SymbolAfterTrue", arg);
+            arg = null;
+            hasValue = arg.HasValue;
+            Tag("HasValueAfterNull", hasValue);
+            Tag("SymbolAfterNull", arg);
+            """;
+        var validator = SETestContext.CreateCS(code, ", bool? arg").Validator;
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs:79]: The test would be easier to read if we use `int?` instead. It's confusing to keep track of `HasValue`, `arg`, and `BoolConstraint` at the same time. Maybe add a second test case with `int?`?

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
It looks like we are on the right track.

I think a second test case would help in understanding what's going on here.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs:79]: `int` doesn't have any constraint on it. I wouldn't mind if you rewrite it in your PRs. As there's ` ToDo` for you here anyway :)

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**